Docs: Invalid return type in middleware example

### What were you searching in the docs?

All examples on [the middleware factory page](https://docs.powertools.aws.dev/lambda/python/latest/utilities/middleware_factory/) present a middleware function with return type of `Callable` which seems incorrect.

<img width="626" alt="image" src="https://github.com/aws-powertools/powertools-lambda-python/assets/22010517/f16a6ad1-9843-407b-83d4-4c0e8039e29f">

The actual return type of the `middleware_before` function is the same as the one of the lambda handler - `dict` this case. Only after the function is decorated, it'll be a `Callable` but I believe the function definition should anyway have the return type of `dict`.

A small reproducible example:
```python3
from typing import Callable

from aws_lambda_powertools.middleware_factory import lambda_handler_decorator

LambdaHandler = Callable[[dict, dict], dict]


@lambda_handler_decorator
def middleware_add_headers(
    handler: LambdaHandler, event: dict, context: dict
) -> Callable:
    return handler(event, context)


@middleware_add_headers
def lambda_handler(event: dict, context: dict) -> dict:
    return {
        "statusCode": 200,
        "body": "hello",
    }
```
Mypy output:
```
$ mypy main.py
handler.py:12: error: Incompatible return value type (got "dict[Any, Any]", expected "Callable[..., Any]")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

### Is this related to an existing documentation section?

https://docs.powertools.aws.dev/lambda/python/latest/utilities/middleware_factory/#middleware-with-before-logic

### How can we improve?

All invalid examples should be updated with correct return types (seems like `dict` in all cases).

### Got a suggestion in mind?

I'm happy to help with resolving this issue.

### Acknowledgment

- [X] I understand the final update might be different from my proposed suggestion, or refused.

--- a/docs/utilities/middleware_factory.md
+++ b/docs/utilities/middleware_factory.md
@@ -30,7 +30,7 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 ### Middleware with before logic
 
 === "getting_started_middleware_before_logic_function.py"
-    ```python hl_lines="5 26 27 32 33 35 40 41"
+    ```python hl_lines="5 26-31 36 37 39 44 45"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_before_logic_function.py"
     ```
 
@@ -43,7 +43,7 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 ### Middleware with after logic
 
 === "getting_started_middleware_after_logic_function.py"
-    ```python hl_lines="7 14 15 21-23 37"
+    ```python hl_lines="8 14-19 24-26 40 41"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_after_logic_function.py"
     ```
 
@@ -58,7 +58,7 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 You can also have your own keyword arguments after the mandatory arguments.
 
 === "getting_started_middleware_with_params_function.py"
-    ```python hl_lines="6 30 31 32 36 52"
+    ```python hl_lines="6 30-37 41 56 57"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_with_params_function.py"
     ```
 
@@ -83,7 +83,7 @@ You can also use [`POWERTOOLS_TRACE_MIDDLEWARES`](#tracing-middleware-execution)
 For advanced use cases, you can instantiate [Tracer](../core/tracer.md){target="_blank"} inside your middleware, and add annotations as well as metadata for additional operational insights.
 
 === "advanced_middleware_tracer_function.py"
-    ```python hl_lines="7 9 12 16 17 19 25 42"
+    ```python hl_lines="7 9 12 16 17 22 28 45 46"
     --8<-- "examples/middleware_factory/src/advanced_middleware_tracer_function.py"
     ```
 
@@ -105,7 +105,7 @@ This makes use of an existing Tracer instance that you may have initialized anyw
     You must [enable Active Tracing](../core/tracer.md#permissions){target="_blank"} in your Lambda function when using this feature, otherwise Lambda cannot send traces to XRay.
 
 === "getting_started_middleware_tracer_function.py"
-    ```python hl_lines="8 14 15 36"
+    ```python hl_lines="8 14 15 39 40"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_tracer_function.py"
     ```
 
@@ -134,7 +134,7 @@ In the example below, we create a Middleware with the following features:
 * Save execution history to a DynamoDB table
 
 === "combining_powertools_utilities_function.py"
-    ```python hl_lines="11 28 29 119 52 61 73"
+    ```python hl_lines="11 28 29 56 64 77 123"
     --8<-- "examples/middleware_factory/src/combining_powertools_utilities_function.py"
     ```
 

--- a/docs/utilities/middleware_factory.md
+++ b/docs/utilities/middleware_factory.md
@@ -30,7 +30,7 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 ### Middleware with before logic
 
 === "getting_started_middleware_before_logic_function.py"
-    ```python hl_lines="5 26-31 36 37 39 44 45"
+    ```python hl_lines="5 26 27 36 37 39 44 45"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_before_logic_function.py"
     ```
 
@@ -43,7 +43,7 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 ### Middleware with after logic
 
 === "getting_started_middleware_after_logic_function.py"
-    ```python hl_lines="8 14-19 24-26 40 41"
+    ```python hl_lines="8 14 15 24-26 40 41"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_after_logic_function.py"
     ```
 
@@ -58,7 +58,7 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 You can also have your own keyword arguments after the mandatory arguments.
 
 === "getting_started_middleware_with_params_function.py"
-    ```python hl_lines="6 30-37 41 56 57"
+    ```python hl_lines="6 30 31 41 56 57"
     --8<-- "examples/middleware_factory/src/getting_started_middleware_with_params_function.py"
     ```
 

--- a/examples/middleware_factory/src/advanced_middleware_tracer_function.py
+++ b/examples/middleware_factory/src/advanced_middleware_tracer_function.py
@@ -14,8 +14,11 @@ app = APIGatewayRestResolver()
 
 
 @lambda_handler_decorator(trace_execution=True)
-def middleware_with_advanced_tracing(handler, event, context) -> Callable:
-
+def middleware_with_advanced_tracing(
+    handler: Callable[[dict, LambdaContext]],
+    event: dict,
+    context: LambdaContext,
+) -> dict:
     tracer.put_metadata(key="resource", value=event.get("resource"))
 
     start_time = time.time()

--- a/examples/middleware_factory/src/advanced_middleware_tracer_function.py
+++ b/examples/middleware_factory/src/advanced_middleware_tracer_function.py
@@ -15,7 +15,7 @@ app = APIGatewayRestResolver()
 
 @lambda_handler_decorator(trace_execution=True)
 def middleware_with_advanced_tracing(
-    handler: Callable[[dict, LambdaContext]],
+    handler: Callable[[dict, LambdaContext], dict],
     event: dict,
     context: LambdaContext,
 ) -> dict:

--- a/examples/middleware_factory/src/combining_powertools_utilities_function.py
+++ b/examples/middleware_factory/src/combining_powertools_utilities_function.py
@@ -26,7 +26,11 @@ feature_flags = FeatureFlags(store=app_config)
 
 
 @lambda_handler_decorator(trace_execution=True)
-def middleware_custom(handler: Callable, event: dict, context: LambdaContext):
+def middleware_custom(
+    handler: Callable[[dict, LambdaContext], dict],
+    event: dict,
+    context: LambdaContext,
+) -> dict:
     # validating the INPUT with the given schema
     # X-Customer-Id header must be informed in all requests
     try:

--- a/examples/middleware_factory/src/getting_started_middleware_after_logic_function.py
+++ b/examples/middleware_factory/src/getting_started_middleware_after_logic_function.py
@@ -12,8 +12,11 @@ app = APIGatewayRestResolver()
 
 
 @lambda_handler_decorator
-def middleware_after(handler, event, context) -> Callable:
-
+def middleware_after(
+    handler: Callable[[dict, LambdaContext], dict],
+    event: dict,
+    context: LambdaContext,
+) -> dict:
     start_time = time.time()
     response = handler(event, context)
     execution_time = time.time() - start_time

--- a/examples/middleware_factory/src/getting_started_middleware_before_logic_function.py
+++ b/examples/middleware_factory/src/getting_started_middleware_before_logic_function.py
@@ -24,7 +24,11 @@ class PaymentError(Exception):
 
 
 @lambda_handler_decorator
-def middleware_before(handler, event, context) -> Callable:
+def middleware_before(
+    handler: Callable[[dict, LambdaContext], dict],
+    event: dict,
+    context: LambdaContext,
+) -> dict:
     # extract payload from a EventBridge event
     detail: dict = extract_data_from_envelope(data=event, envelope=envelopes.EVENTBRIDGE)
 
@@ -38,7 +42,7 @@ def middleware_before(handler, event, context) -> Callable:
 
 
 @middleware_before
-def lambda_handler(event, context: LambdaContext) -> dict:
+def lambda_handler(event: dict, context: LambdaContext) -> dict:
     try:
         payment_payload: dict = extract_data_from_envelope(data=event, envelope=envelopes.EVENTBRIDGE)
         return {

--- a/examples/middleware_factory/src/getting_started_middleware_tracer_function.py
+++ b/examples/middleware_factory/src/getting_started_middleware_tracer_function.py
@@ -12,8 +12,11 @@ app = APIGatewayRestResolver()
 
 
 @lambda_handler_decorator(trace_execution=True)
-def middleware_with_tracing(handler, event, context) -> Callable:
-
+def middleware_with_tracing(
+    handler: Callable[[dict, LambdaContext], dict],
+    event: dict,
+    context: LambdaContext,
+) -> dict:
     start_time = time.time()
     response = handler(event, context)
     execution_time = time.time() - start_time

--- a/examples/middleware_factory/src/getting_started_middleware_with_params_function.py
+++ b/examples/middleware_factory/src/getting_started_middleware_with_params_function.py
@@ -28,7 +28,12 @@ class BookingError(Exception):
 
 
 @lambda_handler_decorator
-def obfuscate_sensitive_data(handler, event, context, fields: List) -> Callable:
+def obfuscate_sensitive_data(
+    handler: Callable[[dict, LambdaContext], dict],
+    event: dict,
+    context: LambdaContext,
+    fields: List,
+) -> dict:
     # extracting payload from a EventBridge event
     detail: dict = extract_data_from_envelope(data=event, envelope=envelopes.EVENTBRIDGE)
     guest_data: Any = detail.get("guest")
@@ -49,7 +54,7 @@ def obfuscate_data(value: str) -> bytes:
 
 
 @obfuscate_sensitive_data(fields=["email", "passport", "vat"])
-def lambda_handler(event, context: LambdaContext) -> dict:
+def lambda_handler(event: dict, context: LambdaContext) -> dict:
     try:
         booking_payload: dict = extract_data_from_envelope(data=event, envelope=envelopes.EVENTBRIDGE)
         return {

